utils_disk: Add disk resize related functions

peixiu opened the pull request:
For commits utils_disk: Add disk resize related functions:
1. Add get_partition_attrs_linux() to get a disk
   partition's attributes in linux geust.
2. Add get_mpoint_fstype_linux() to get a disk
   partition's file system type and mountpoint.
3. Add resize_filesystem_linux() to resize a partition's
   filesystem in linux guest.
4. Add resize_partition_linux() to resize a disk
   partition's size in linux guest.
5. add get_disk_size_windows() to get a disk size in windows guest.
6. add get_disk_size_linux() to get a disk size in linux guest.
7. add get_disk_size() to get a disk size in guest.

id:1664992
Signed-off-by: Peixiu Hou <phou@redhat.com>

XueqiangWei commented:
Conflicted with avocado-vt  #1908, please contact yhong

xiagao commented on virttest/utils_disk.py:
```diff
@@ -747,7 +833,9 @@ def configure_empty_linux_disk(session, did, size, start="0M", n_partitions=1,
                                        str(start) + size[-1], part_type, timeout)
         start += partition_size
         post_partition = get_linux_disks(session, partition=True).keys()
+        print ("post_partition is %s" % post_partition)
```
why not use logging.info

peixiu replied:
oo, it's debug info, will delete them, thanks

xiagao commented on virttest/utils_disk.py:
```diff
@@ -747,7 +833,9 @@ def configure_empty_linux_disk(session, did, size, start="0M", n_partitions=1,
         post_partition = get_linux_disks(session, partition=True).keys()
+        print ("post_partition is %s" % post_partition)
         new_partition = list(set(post_partition) - set(pre_partition))[0]
+        print ("new_partition is %s" % new_partition)
```
the same

peixiu replied:
Updated, thanks

peixiu commented:
@xiagao @XueqiangWei @YongxueHong Could you help to review this patch? 
For the CI check failed reason is not  related this patch, but I don't know how to solve it~
For will conflict with avocado-vt #1908 issue, has confirmed with yhong, no matter which patch merged first, another one need to solve the conflict issue, but now, it won't block the code review.
so, above all, could please review the code first? 

thanks a lot~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,90 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
         create_filesyetem_linux(session, partition_name, fstype, timeout)
 
 
+def resize_xfs_filesystem(session, partition, fstype='xfs', size=None):
```
Hi lolyu, 
thanks for your review, session.cmd also can catch nozero cmd status, it will raise ShellCmdError directly, and will return the output of cmd, I think it's enough for use to know what happen when error occur~ so no use session.cmd_status() here.

xiagao commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,90 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+
+
+def resize_partition_linux(session, dev_name, partition,
+                           size, fstype='ext4', start='0M'):
```
If dev_name can be got from partition,if yes, it will reduce a variable

peixiu commented:
Tested with latest codes:
 (1/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (618.77 s)
 (2/2) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_md5_check.with_data_disk.fmt_qcow2.q35: PASS (422.68 s)

luckyh commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,88 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
         create_filesyetem_linux(session, partition_name, fstype, timeout)
 
 
+def resize_xfs_filesystem(session, partition, fstype='xfs', size=None):
```
Why do we need an argument for `fstype`?

luckyh replied:
The `size` can be a positive or negative number, otherwise, this function is only used for 'enlarging', so the default should be `0` instead of `None`, or it doesn't make sense to have a default value.

luckyh commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,88 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :return mpoint: return mountpoint.
+    """
+    mpoint = '/mnt/' + utils_misc.generate_random_string(5)
+    session.cmd('mkdir {0} && mount {1} {0}'.format(mpoint, partition))
```
This should not be done within the function.

peixiu commented:
@luckyh thanks for your advice, updated the code~
Changed follows:
1.  United resize_xfs_filesystem() and resize_ext_filesystem() to resize_filesystem_linux() function.
2.  United arguments to (session, partition, size).
3. In order to get fstype and mountpoint, add a new function get_fstype_mpoint_linux() function.

Please help to review it again~
Thanks a lot~

peixiu commented:
Test with latest codes:
 (1/2) Host_RHEL.m*.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_md5_check.with_data_disk.fmt_qcow2.q35: PASS (429.53 s)
 (2/2) Host_RHEL.m*.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.block_resize.xfs_fstype.data_plane.with_md5_check.with_data_disk.fmt_qcow2.q35: PASS (302.37 s)

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,96 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    session.cmd(rm_part_cmd)
+
+    size = float(utils_misc.normalize_data_size(size, order_magnitude="M",
+                                                factor="1000"))
```
Why set the argument "factor" to "1000" by force?

peixiu replied:
Due to used parted command here, it's size unit is 1000, not 1024~

peixiu replied:
Changed to use Mib, thanks~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,96 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+
+    size = float(utils_misc.normalize_data_size(size, order_magnitude="M",
+                                                factor="1000"))
+    resizepart_cmd = ' '.join((parted_cmd, 'mkpart primary {0} {1}M'))
```
How to confirm the part type is "primary"?

peixiu replied:
updated the code, get the partition type first~
thanks~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,96 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        if size == 'MAX_AVAILABLE_SIZE':
+            resize_fs_cmd = 'xfs_growfs -d %s' % mountpoint
+        else:
+            resize_fs_cms = 'xfs_growfs -D %s %s' % (size, mountpoint)
```
1. will update line 640 to "resize_fs_cmd", thanks~
2. for xfs file system resize only support enlarge, not support shrink, it's depends on xfs specify feature, I will add an explanation for xfs file system.
thanks~

peixiu replied:
updated thanks~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -24,6 +24,7 @@
 
 from virttest import error_context
 from virttest import utils_numeric
+from virttest import utils_misc
```
updated, thanks

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +601,96 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        return (fstype, mountpoint)
+    else:
+        logging.error('partition should not be mounted, mounted'
+                      'list as:\n%s' % mount_list)
```
updated, thanks

peixiu commented:
@xiagao @XueqiangWei @YongxueHong 
Could you help to review this patch again?
Thanks a lot~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +600,103 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        mount(partition, mountpoint, fstype=fstype, session=session)
+
+
+def resize_partition_linux(session, partition, size, start='0M'):
```
Some errors:
```
>>> print(session.cmd('lsblk'))
NAME                      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda                         8:0    0   20G  0 disk 
├─sda1                      8:1    0    4M  0 part 
├─sda2                      8:2    0    1G  0 part /boot
└─sda3                      8:3    0   19G  0 part 
  ├─rhel_dhcp71--240-root 253:0    0   17G  0 lvm  /
  └─rhel_dhcp71--240-swap 253:1    0    2G  0 lvm  [SWAP]
sdb                         8:16   0  100G  0 disk 
└─sdb1                      8:17   0 94.4G  0 part /mnt/sdb1

>>> utils_disk.resize_partition_linux(session, '/dev/sdb1', '5000M', start='0M')
Traceback (most recent call last):
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1130, in cmd_output
    o = self.read_up_to_prompt(timeout, internal_timeout, print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1099, in read_up_to_prompt
    print_func)[1]
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 925, in read_until_last_line_matches
    print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 843, in read_until_output_matches
    raise ExpectTimeoutError(patterns, o)
aexpect.exceptions.ExpectTimeoutError: Timeout expired while looking for pattern '^\\[.*\\][\\#\\$]\\s*$'    (output: 'e2fsck 1.44.3 (10-July-2018)\nError reading block 12091392 (Invalid argument).  Ignore error<y>? ')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yhong/kar/workspace/avocado-vt/virttest/utils_disk.py", line 695, in resize_partition_linux
    session.cmd('e2fsck -f %s' % partition)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1286, in cmd
    print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1213, in cmd_status_output
    o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1134, in cmd_output
    raise ShellTimeoutError(cmd, o)
aexpect.exceptions.ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'e2fsck -f /dev/sdb1'    (output: 'e2fsck 1.44.3 (10-July-2018)\nError reading block 12091392 (Invalid argument).  Ignore error<y>? ')
>>> 

```

peixiu replied:
As we discussed, this function is connection with  resize_filesystem_linux(), in order to reduce user misunderstand this connection, so I packaged enlarge_disk_size_linux() and shrink_disk_size_linux() functions as well.

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +600,103 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+                  'list as:\n%s' % mount_list)
+
+
+def resize_filesystem_linux(session, partition, size):
```
Updated, thanks~

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+                 enlarge to maximun available size.
+    """
+    fstype, mountpoint = get_fstype_mpoint_linux(session, partition)
+    if fstype == 'xfs':
```
Suggest to add comments to doc string.  xfs doesn't support shrink.

peixiu replied:
that's ok, will updated~

peixiu commented:
@XueqiangWei @YongxueHong @xiagao I updated the code, could you help to review this patch again?
Thanks a lot~

peixiu commented:
@CongLi could you help to review this patch? thanks~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,155 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        else:
+            get_bsize_cmd = 'xfs_growfs -n %s' % mountpoint
+            output = session.cmd_output(get_bsize_cmd)
+            bsize = re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0]
```
cannot use re.match here due to the data** characters don't at the beginning of the string, we can use re.search, but it also need use .group() to pick the value out, the result same as findall, I used accurate pattern, only have 1 correct value in output.

> Suggest use the match interface.
> refer to python.org(https://docs.python.org/3/library/re.html?highlight=re#module-re):
> 
> > findall() matches all occurrences of a pattern, not just the first one as search() does. For example, if a writer wanted to find all of the adverbs in some text, they might use findall() in the following manner:
> 
> https://docs.python.org/3/library/re.html?highlight=re#search-vs-match
> 
> > search() vs. match()
> > Python offers two different primitive operations based on regular expressions: re.match() checks for a match only at the beginning of the string, while re.search() checks for a match anywhere in the string (this is what Perl does by default).

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,155 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+            get_bsize_cmd = 'xfs_growfs -n %s' % mountpoint
+            output = session.cmd_output(get_bsize_cmd)
+            bsize = re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0]
+            blocks = int(size.split(".")[0]) / int(bsize) - 512
```
updated

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,155 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    """
+    resize_filesystem_linux(session, partition, size)
+    resize_partition_linux(session, partition, size)
+
```
updated as we discussed with xuhan.

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,155 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    resize_partition_linux(session, partition, size)
+
+
+def get_disk_size(session, os_type, did):
```
updated as we discussed with xuhan.

peixiu commented:
Tested with rhel8.0.0+ext4 and rhel8.0.0+xfs and windows2016, all passed.

PASS 1-Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw
PASS 1-Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw
PASS 1-Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.with_xfs.with_iozone.with_data_disk.fmt_raw

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,155 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :param size: resize partition to size, unit is B.
+    """
+    resize_partition_linux(session, partition, size)
+    resize_filesystem_linux(session, partition, AVAILABLE_SIZE)
```
has moved to case py file, after the partition has been enlarged, the filesystem can be enlarge to maximum available_size, so here use this param.

peixiu commented:
@YongxueHong @luckyh @XueqiangWei I updated the code, could you help to review it again? thanks a lot~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+            get_bsize_cmd = 'xfs_growfs -n %s' % mountpoint
+            output = session.cmd_output(get_bsize_cmd)
+            bsize = re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0]
+            blocks = int(size.split(".")[0]) // int(bsize) - 512
```
How do you confirm to need to subtract 512 blocks in `blocks = int(size.split(".")[0]) // int(bsize) - 512` , not 1024 blocks or 4096 blocks. Is it related with the options "physical_block_size" and "logical_block_size" ?
e.g:
```
-drive id=drive_disk1,if=none,snapshot=off,aio=native,cache=none,format=qcow2,file=/root/test_home/yhong/images/storage.qcow2 \
-device scsi-hd,id=disk1,drive=drive_disk1,physical_block_size=4096,logical_block_size=1024 \
```

peixiu replied:
Has updated, thanks~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+            resize_fs_cmd = 'resize2fs %s' % partition
+        else:
+            size = utils_numeric.normalize_data_size(size, order_magnitude="M")
+            size = int(size.split(".")[0]) - 2
```
Also here, How to confirm to subtract 2M?

peixiu replied:
Has updated to size - start_size, thanks~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+            resize_fs_cmd = 'resize2fs %s' % partition
+        else:
+            size = utils_numeric.normalize_data_size(size, order_magnitude="M")
+            size = int(size.split(".")[0]) - 2
```
I think there is a calculation error for `size.split(".")[0]`.
If the size is "100.9M" after `size = utils_numeric.normalize_data_size(size, order_magnitude="M")`, 
 then the size is '100M' after `int(size.split(".")[0])`. It will igore about 1M.
So, it may be not accurate.

peixiu replied:
for filesystem resize, ignore 0.9M is can be acceptance, it's size need smaller then the partition size and need auto alignment~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        mount(partition, mountpoint, fstype=fstype, session=session)
+
+
+def resize_partition_linux(session, partition, size, start='0M'):
```
I think that doesn't need to pass the argument `start=0`, just `def resize_partition_linux(session, partition, size):`

peixiu replied:
tried to set the start value in  function, thanks~

peixiu replied:
@YongxueHong 
updated the code, please help to review it again~ 
thanks a lot~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    if os_type == "linux":
+        return get_disk_size_linux(session, did)
+    else:
+        return get_disk_size_windows(session, did)
```
Updated, thanks~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,154 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        return get_disk_size_linux(session, did)
+    else:
+        return get_disk_size_windows(session, did)
+
```
updated~

peixiu commented:
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (254.05 s)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.xfs_fstype.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (104.86 s)
 (1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (205.22 s)

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,177 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    part_info = re.match(r'(/dev/.*)(\d+)', partition)
+    dev_name, part_num = part_info.groups()
+
+    parted_cmd = 'parted -s %s print' % dev_name
```
that's ok, will update~

peixiu replied:
Updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,177 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :param session: session object to guest.
+    :param partition: partition that to be shrunk, like /dev/sdb1.
+    :param size: resize partition to size, unit is B.
+    :param start: mkpart primary start place, default is 0M.
```
that's ok, will update

peixiu replied:
updated

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,177 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    pattern = r"%s\s+(\d+)" % did
+    output = session.cmd_output(cmd)
+    disk_size = re.findall(pattern, output)[0]
+    return float(disk_size)
```
ok, thanks, will update

peixiu replied:
updated

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,177 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        mount(partition, mountpoint, fstype=fstype, session=session)
+
+
+def get_start_type_linux(session, partition):
```
how about get_partition_st_linux()? s for start size, t for type?

peixiu replied:
updated, thanks

peixiu commented on virttest/utils_disk.py:
```diff
@@ -600,6 +602,177 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        mountpoint = mount_info.group(1)
+        fstype = mount_info.group(2)
+        return (fstype, mountpoint)
+    logging.error('partition is not mounted, mounted'
```
it's ok, thanks

peixiu replied:
updated

peixiu commented:
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (199.93 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.xfs_fstype.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (110.86 s)

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,179 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    mount_info = re.search(pattern, mount_list)
+    if mount_info:
+        return mount_info.groups()
+    else:
```
updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,179 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    pattern = re.compile(r'%s\s+(?P<start>\d+\w+)\s+(?P<end>\d+\w+)\s+'
+                         r'(?P<size>\d+\w+)\s+(?P<type>\w+)' % part_num)
+    part_attrs = pattern.search(output, re.M).groupdict()
+    return part_attrs
```
updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,179 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        if is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+            umount(partition, mountpoint, fstype=fstype, session=session)
+
+        session.cmd('e2fsck -f %s' % partition)
```
It's a e2fsck check step, if cannot pass, won't execute next~

peixiu commented:
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (214.99 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.xfs_fstype.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (111.60 s)

peixiu commented:
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (203.81 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.0.x86_64.io-github-autotest-qemu.block_resize.xfs_fstype.data_plane.with_iozone.with_data_disk.fmt_raw: PASS (116.36 s)

YongxueHong commented:
For the following functions:
add get_disk_size_windows()
add get_disk_size_linux()
add get_disk_size()

They are not related to the functions of resizing, pls new another commit for them on this pull request.
Thanks.

YongxueHong commented:
The commit is not detailed enough, pls add a summary to each function.

peixiu commented:
@YongxueHong updated the commit info, please help to review again~
Thanks

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,163 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :return: dict like {'start': '512B', 'end': '16106127359B',
+                       'size': '16106126848B', 'type': 'primary'}
+    """
+    dev_name, part_num = re.match(r'(/dev/.*)(\d+)', partition).groups()
```
Should consider the NVMe type disk.
e.g:
```
[root@ibm-p8-garrison-04 ~]# lsblk
NAME                                MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                                   8:0    1 931.5G  0 disk 
├─sda1                                8:1    1     4M  0 part 
├─sda2                                8:2    1     1G  0 part /boot
└─sda3                                8:3    1 930.5G  0 part 
  ├─rhel_ibm--p8--garrison--04-root 253:0    0    50G  0 lvm  /
  └─rhel_ibm--p8--garrison--04-home 253:2    0   3.2T  0 lvm  /home
sdb                                   8:16   1 931.5G  0 disk 
└─sdb1                                8:17   1 931.5G  0 part 
  └─rhel_ibm--p8--garrison--04-home 253:2    0   3.2T  0 lvm  /home
sr0                                  11:0    1  1024M  0 rom  
sr1                                  11:1    1  1024M  0 rom  
sr2                                  11:2    1  1024M  0 rom  
sr3                                  11:3    1  1024M  0 rom  
nvme0n1                             259:0    0   1.5T  0 disk 
└─nvme0n1p1                         259:1    0   1.5T  0 part 
  ├─rhel_ibm--p8--garrison--04-swap 253:1    0     4G  0 lvm  [SWAP]
  └─rhel_ibm--p8--garrison--04-home 253:2    0   3.2T  0 lvm  /home

[root@ibm-p8-garrison-04 ~]# parted /dev/nvme0n1
GNU Parted 3.1
Using /dev/nvme0n1
Welcome to GNU Parted! Type 'help' to view a list of commands.
(parted) p                                                                
Model: NVMe Device (nvme)
Disk /dev/nvme0n1: 1600GB
Sector size (logical/physical): 4096B/4096B
Partition Table: msdos
Disk Flags: 

Number  Start   End     Size    Type     File system  Flags
 1      1049kB  1600GB  1600GB  primary               lvm

(parted)           
```

peixiu replied:
updated~ please help review again, thanks~

YongxueHong commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,167 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    if "nvme" in partition:
+        pattern = r'(/dev/.*)p(\d+)'
+    else:
+        pattern = r'(/dev/.*)(\d+)'
```
Better to `pattern = r'(/dev/.*)p(\d+)' if "nvme" in partition else r'(/dev/.*)(\d+)'` it's more pythonic  :)

peixiu replied:
it's ok~

YongxueHong commented:
LGTM

peixiu commented:
@xueqiangWei @luckyh could you help to review this patch? thanks a lot~

XueqiangWei commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,164 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    """
+    Resize file system in linux guest.
+    For ext2, ext3, ext4 filesystem, support enlarge and shrink.
+    For xfs filesystem, only support enlarge, not support shrink.
```
As we discussed before, not support gpt for linux. Suggest to comment it. In order to prevent  calling incorrectly

peixiu replied:
that's ok, will update~

peixiu replied:
Updated, thanks~

XueqiangWei commented:
Failed case: 3
08:44:56 ERROR| aexpect.exceptions.ShellTimeoutError: Timeout expired while waiting for shell command to complete: '/home/iozone_inst/src/current/iozone  -azR -r 64k -n 512M -g 1G -I -f /mnt/sdb1/testfile'

CaseName                                 Status          StartTime       EndTime         Result     TimeElapsed
'01-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-17 08:07:05 2019-07-17 08:44:56 ERROR 2271.289204
'02-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-17 08:44:56 2019-07-17 09:20:20 ERROR 2123.645931
'03-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-17 09:20:20 2019-07-17 09:58:41 ERROR 2300.875202
'04-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-17 09:58:41 2019-07-17 11:04:13 PASS 3931.734
'05-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-17 11:04:13 2019-07-17 11:28:28 PASS 1455.253441
'06-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-17 11:28:28 2019-07-17 11:49:19 PASS 1250.616407
'07-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_system_disk.q35' Finshed 2019-07-17 11:49:19 2019-07-17 12:40:22 PASS 3063.438793
'08-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-17 12:40:22 2019-07-17 13:04:01 PASS 1418.87502
'09-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-17 13:04:01 2019-07-17 13:25:37 PASS 1295.430917
'10-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_system_disk.q35' Finshed 2019-07-17 13:25:37 2019-07-17 14:05:18 PASS 2380.761243
Summary: 
Finshed=10, ERROR=3, PASS=7

XueqiangWei commented:
Tested related case: all passed.
CaseName Status StartTime EndTime Result TimeElapsed
'01-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-19 06:24:18 2019-07-19 06:44:53 PASS 1235.097967
'02-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-19 06:44:53 2019-07-19 07:03:22 PASS 1108.497018
'03-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-19 07:03:22 2019-07-19 07:23:46 PASS 1224.84459
'04-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-19 07:23:47 2019-07-19 07:42:15 PASS 1108.577616
'05-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-19 07:42:15 2019-07-19 08:00:04 PASS 1069.19236
'06-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-19 08:00:04 2019-07-19 08:15:32 PASS 927.694781
'07-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_system_disk.q35' Finshed 2019-07-19 08:15:32 2019-07-19 09:09:47 PASS 3255.122622
'08-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35' Finshed 2019-07-19 09:09:47 2019-07-19 09:27:33 PASS 1065.562417
'09-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35' Finshed 2019-07-19 09:27:33 2019-07-19 09:43:20 PASS 947.446277
'10-Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_system_disk.q35' Finshed 2019-07-19 09:43:20 2019-07-19 10:22:39 PASS 2358.982745
Summary:
Finshed=10, PASS=10

LGTM,
Acked-by xuwei@redhat.com

peixiu commented:
@luckyh could you help to review this patch? thanks a lot~

luckyh commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
         create_filesyetem_linux(session, partition_name, fstype, timeout)
 
 
+def get_mpoint_fstype_linux(session, partition):
```
I think you want `get_fs_attrs_linux()`, otherwise, it should be a private function if you just want some of the attrs.

peixiu replied:
thanks, updated~

luckyh commented on virttest/utils_disk.py:
```diff
@@ -33,6 +33,8 @@
 PARTITION_TYPE_EXTENDED = "extended"
 PARTITION_TYPE_LOGICAL = "logical"
 
+AVAILABLE_SIZE = "available size"
```
Suggest the name to be `SIZE_AVAILABLE`

peixiu replied:
thanks, updated~

luckyh commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    if mount_info:
+        return mount_info.groups()
+    raise exceptions.TestError('partition is not mounted, mounted'
+                               'list as:\n%s' % mount_list)
```
Library functions should return a negative value instead of commit a `TestError` when the result is unexpected.

peixiu replied:
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    mountpoint, fstype = get_mpoint_fstype_linux(session, partition)
+    if fstype == 'xfs':
+        resize_xfs_fs(size)
+    else:
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        resize_xfs_fs(size)
+    else:
+        resize_ext_fs(size)
+        if not is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+        else:
+            output = session.cmd_output('xfs_growfs -n %s' % mountpoint)
+            bsize = int(re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0])
+            blocks = (float(utils_numeric.normalize_data_size(size, 'B')) -
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    if is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+        umount(partition, mountpoint, fstype=fstype, session=session)
+
+    dev_name, part_num = re.match(r'(/dev/.*)(\d+)', partition).groups()
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    :return: disk size.
+    """
+    cmd = "wmic diskdrive get size, index"
+    return float(re.findall(r'%s\s+(\d+)' % did, session.cmd_output(cmd))[0])
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+    """
+    disks_info = get_linux_disks(session, partition=True)
+    disk_size = disks_info['%s' % did][1]
+    return float(utils_numeric.normalize_data_size(disk_size, 'B'))
```
thanks, updated~

peixiu commented on virttest/utils_disk.py:
```diff
@@ -601,6 +603,165 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
+
+    if fstype == 'xfs':
+        session.cmd('xfs_repair -n %s' % partition)
+    else:
```
thanks, updated~

peixiu commented:
(1/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (238.31 s)
 (2/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35: PASS (203.50 s)
 (3/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_system_disk.q35: PASS (553.67 s)
 (4/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (247.49 s)
 (5/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35: PASS (208.59 s)
 (6/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_system_disk.q35: PASS (346.76 s)

 (1/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (186.07 s)
 (2/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35: PASS (191.07 s)
 (3/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (189.88 s)
 (4/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35: PASS (193.38 s)

peixiu commented:
tested with latest code:
 (1/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (188.90 s)
 (2/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35: PASS (189.82 s)
 (3/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (252.39 s)
 (4/4) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35: PASS (174.96 s)

 (1/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (233.70 s)
 (2/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_data_disk.fmt_raw.q35: PASS (217.53 s)
 (3/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.with_iozone.with_system_disk.q35: PASS (665.41 s)
 (4/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_qcow2.q35: PASS (275.94 s)
 (5/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_data_disk.fmt_raw.q35: PASS (225.77 s)
 (6/6) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_resize.data_plane.with_iozone.with_system_disk.q35: PASS (433.21 s)

luckyh commented:
Thanks all.
